zerver/tornado/event_queue.py

```diff
@@ -1236,7 +1235,47 @@ def process_notification(notice: Mapping[str, Any]) -> None:
     start_time = time.time()
 
     if event["type"] == "message":
-        process_message_event(event, cast(List[Mapping[str, Any]], users))
+        if len(users) > 0 and isinstance(users[0], dict) and "stream_push_notify" in users[0]:
+            # TODO/compatibility: Remove this whole `if` block once one can no
+            # longer directly upgrade directly from 4.x to master.
+            # do_send_messages used to send events with users in dict format, with the
+            # dict containing the user_id and other data. We later trimmed down the user
+            # data to only contain the user_id and flags, and put everything else in
+            # the event dict as lists.
+            # This block handles the old-format events still in the queue during upgrade.
+
+            modern_event = cast(MutableMapping[str, Any], event)
+            user_dicts = cast(List[MutableMapping[str, Any]], users)
+
+            # Back-calculate the older all-flags format data in the `users` dicts into the newer
+            # all-lists format, and attach the lists to the `event` object.
+            modern_event["online_push_user_ids"] = []
+            modern_event["stream_push_user_ids"] = []
+            modern_event["stream_email_user_ids"] = []
+            modern_event["wildcard_mention_user_ids"] = []
+            modern_event["muted_sender_user_ids"] = []
+
+            for user in user_dicts:
+                user_id = user["id"]
+
+                if user.pop("stream_push_notify", False):
+                    modern_event["stream_push_user_ids"].append(user_id)
+                if user.pop("stream_email_notify", False):
+                    modern_event["stream_email_user_ids"].append(user_id)
+                if user.pop("wildcard_mention_notify", False):
+                    modern_event["wildcard_mention_user_ids"].append(user_id)
+                if user.pop("sender_is_muted", False):
+                    modern_event["muted_sender_user_ids"].append(user_id)
+
+                # TODO/compatibility: Another translation code block for the rename of
+                # `always_push_notify` to `online_push_enabled`.  Remove this
+                # when one can no longer directly upgrade from 4.x to master.
+                if user.get("online_push_enabled", False) or user.get("always_push_notify", False):
```

Comment:
Should these be `user.pop()`?  

I'd also propose that we move this block of code to a function, `reformat_legacy_send_message_event`, and then we can write a couple simple tests for it (you can generate dictionaries for testing it by printing before/after events), in addition to keeping it out of the main flow for this function.

Reply:
Done, thanks!